### ISSUE TITLE
[GSD-13038] Regression in 26.09+: dual-GPU workloads eagerly pin a host-RAM mirror of every large device USM allocation on the peer device, exhausting system memory

### ISSUE THREAD

**Issue by blanelegant** [intel/compute-runtime#953]
## Summary

On a 2× Arc Pro B70 (Battlemage G31) system running PyTorch DDP (one rank per card,
xccl/oneCCL collectives), every large device-USM segment a rank allocates on its own
card is also registered on its file descriptor to the **peer** card as GTT (GPU-mapped
system memory). On 26.05 these peer registrations are bookkeeping only. On 26.09+
(observed on 26.22) they are **eagerly pinned to physical host pages**: ~17.7 GiB of
system RAM per rank for an ~18.5 GiB VRAM working set — 35.4 GiB total on a 64 GiB
machine. The kernel then hits order-0 page-allocation failures, the desktop swaps out,
and at production scale the workload dies. Downgrading only `intel-compute-runtime`
26.22 → 26.05 (nothing else changed) reduces pinned memory for the identical workload
from **25.9 GiB to 0.67 GiB** with identical throughput.

## Environment

| component | version |
|---|---|
| intel-compute-runtime | **26.22.38646.4 (bad)** / **26.05.37020.3 (good)** |
| GPUs | 2× Intel Arc Pro B70 (Battlemage G31, 8086:e223), 32 GiB VRAM each, PCIe (no XeLink) |
| kernel / driver | Linux 7.1.3 (CachyOS), `xe` |
| level-zero-loader | 1.28.6 |
| intel-graphics-compiler | 2.36.3 |
| intel-gmmlib | 22.10.0 |
| workload stack | PyTorch 2.13.0+xpu, oneCCL 2022.0 (xccl backend), torchrun 2 ranks |
| system RAM | 64 GiB |

## Reproducer

```python
# gtt_repro.py — torchrun --standalone --nproc-per-node=2 gtt_repro.py
import os, time
import torch
import torch.distributed as dist

rank = int(os.environ["RANK"])
local = int(os.environ["LOCAL_RANK"])
dev = torch.device(f"xpu:{local}")
dist.init_process_group("xccl", device_id=dev)

# 8 GiB of device USM on this rank's card, in 1 GiB segments
held = [torch.empty(1024 ** 3 // 4, dtype=torch.float32, device=dev) for _ in range(8)]
# gradient-bucket-sized collective traffic
bucket = torch.ones(25_000_000, device=dev)
for _ in range(50):
    dist.all_reduce(bucket)
torch.xpu.synchronize(dev)

print(f"rank {rank} pid {os.getpid()} holding 8 GiB on {dev}; sleeping for inspection", flush=True)
time.sleep(120)
dist.barrier()
dist.destroy_process_group()
```

Inspect while it sleeps:

```sh
for pid in <rank pids>; do
  for fd in /proc/$pid/fdinfo/*; do grep -HE 'drm-resident-(gtt|vram0)' $fd; done
done
free -h   # or a kernel with pinned-GPU-memory accounting in /proc/meminfo
```

Observed per-fd accounting for the reproducer (each rank holds 8 GiB on its own card;
note the ~8.9 GiB **GTT on the fd to the peer card**, mirror-imaged between ranks):

```
pid 116400 (rank 0, computes on renderD128)
  fd4 /dev/dri/renderD129: drm-resident-gtt: 8913000 KiB  drm-resident-vram0:  191268 KiB
  fd5 /dev/dri/renderD128: drm-resident-gtt:  642792 KiB  drm-resident-vram0: 9355696 KiB
pid 116401 (rank 1, computes on renderD129)
  fd4 /dev/dri/renderD129: drm-resident-gtt:  638440 KiB  drm-resident-vram0: 9348024 KiB
  fd5 /dev/dri/renderD128: drm-resident-gtt: 9011816 KiB  drm-resident-vram0:  392456 KiB
```

(The capture above is from 26.05, where the registrations exist but stay unpinned.
On 26.22 the same registrations are backed by pinned host pages — see below.)

## Bad vs good, identical real workload (PPO trainer, ~16 GiB VRAM working set per rank)

| | 26.22 | 26.05 |
|---|---|---|
| peer-fd `drm-resident-gtt` per rank | 12.8 GiB | 13.3 GiB (bookkeeping only) |
| pinned host memory, both ranks (kernel `GPUActive`) | **25.9 GiB** | **0.67 GiB** |
| throughput | ~3.0 s/wave | ~3.0 s/wave |

(`GPUActive` is the CachyOS kernel's pinned-GPU-page accounting in `/proc/meminfo`;
on a mainline kernel the same effect is visible as "used" memory in `free` that no
process RSS accounts for, disappearing when the workload exits.)

At full production scale on 26.22 the pinning reached 35.4 GiB (17.7 GiB per rank),
`free` showed ~7 GiB available of 64 GiB, the kernel logged order-0 allocation
failures in unrelated processes (`kwin_wayland: page allocation failure: order:0`),
and the training run died without OOM-killer involvement.

Growth is stepwise and exactly tracks the allocator's large segments: peer-fd GTT grew
in increments of 3,174,400 KiB — precisely the workload's 512×2048×795 fp32 buffer —
and plateaus at ~the VRAM working set once the caching allocator stops creating new
segments. Nothing in the workload allocates host or shared USM of this size; all
tensors are device-resident.

## What I ruled out

- **USM pooling knobs**: `NEOReadDebugKeys=1 EnableHostUsmAllocationPool=0
  EnableDeviceUsmAllocationPool=0 EnableUsmAllocationPoolManager=0` on 26.22 → peer-fd
  GTT unchanged; additionally the workload then crashes in backward with
  `UR_RESULT_ERROR_OUT_OF_RESOURCES` (possibly its own issue).
- **Workload-side caching**: releasing unused allocator segments
  (`torch.xpu.empty_cache()`) after warmup reclaims only the warmup-transient portion
  (~2 GiB/rank); live segments remain mirrored and pinned.
- **VRAM oversubscription/eviction**: `drm-resident-vram0` ≈ `drm-total-vram0`
  throughout; VRAM is never oversubscribed (≤ 19 of 32 GiB).

## Expected behavior

Cross-device registrations for peer access (if that is what these are) should not
eagerly pin physical host pages proportional to the peer's entire VRAM working set —
26.05's lazy behavior, which performs identically for this workload, demonstrates the
pinning is unnecessary.

## Possibly related

- **#952 looks like this same defect presenting as a crash at 4-GPU scale.** There:
  4× B60, 60 GB RAM, ~11.4 GiB reserved per rank, `UR_RESULT_ERROR_OUT_OF_RESOURCES`
  at the first large kernel — exactly what eager peer-pinning predicts (each rank's
  working set mirrored per peer exceeds host RAM at 4 ranks; our 2-GPU/64 GiB system
  survives at 35.4 GiB pinned but dies at the system level instead). Their
  discriminating evidence (single-GPU OK, independent concurrent jobs OK, small-model
  DDP OK, large-model DDP fails on 2/3/4 ranks) matches peer-mirror scaling, and their
  USM-pool debug-key attempts not helping matches ours.
- #935: on these cards (BMG, cross-root-port) L0 dev-to-dev P2P is refused by the
  kernel, so peer access is host-backed — presumably why these peer GTT registrations
  exist at all. The regression here is only that 26.09+ pins them eagerly.
- The 25.40→26.09 window matches the `useUsmPoolManager` / "l0 device usm growing
  pools" default flip referenced in #916 (also a dual-Arc, UR multi-device-context
  setup), though disabling those pools via debug keys did not change the behavior
  here, so the overlap may be coincidental.

**Comment by kgibala** [comment]
Hello @blanelegant,

Thank you for submitting this issue. We will move it internally for further isolation and investigation.

However, we'd like to kindly point out that this issue does not fully comply with our Issue Submission Guide. Please note that in the future, issues that do not follow the submission guidelines may need to be rejected - every issue should respect the guide to ensure efficient triage and resolution.

Could you please follow the [Issue Submission Guide](https://github.com/intel/compute-runtime/blob/master/documentation/ISSUE_SUBMISSION_GUIDE.md) and, going forward, make sure new issues are created based on it?

We appreciate your understanding and cooperation.

**Comment by jwarchul** [comment]
Hi @blanelegant,

We are investigating this regression and suspect that "enable defer backing for bmg" might have caused it. Could you please check if running the test with the flags `NEOReadDebugKeys=1 EnableDeferBacking=0` resolves the issue on your side?

Thanks for your help!

**Comment by blanelegant** [comment]
Hi @jwarchul — confirmed: `EnableDeferBacking=0` resolves it.

Ran the reproducer from the issue body (2 ranks, 8 GiB device USM each +
all_reduce traffic) on **26.22.38646.4**, measuring peak pinned GPU pages
(`GPUActive` in `/proc/meminfo`, delta over idle baseline):

| configuration | peak pinned host pages |
|---|---|
| 26.22 default | **17,457 MiB** |
| 26.22 `NEOReadDebugKeys=1 EnableDeferBacking=0` | **285 MiB** |

The flagged run matches 26.05 behavior (~0.67 GiB pinned for the real
workload in the original report). So "enable defer backing for bmg" does
look like the cause.

Happy to test a fix when one lands. In the meantime, is
`EnableDeferBacking=0` safe to run in production on 26.2x, or are there
known side effects we should weigh against staying pinned on 26.05?

**Comment by jwarchul** [comment]
Hi @blanelegant,

Thanks for confirming it.

Yes, using `EnableDeferBacking=0` should be safe for production. This was essentially the default behavior in version 26.05 and earlier releases, so disabling it restores that previous behavior.

We are working on a proper fix and will update you here as soon as it's ready.

**Comment by jwarchul** [comment]
Hi @blanelegant,

Following further internal investigation, debug keys (such as `EnableDeferBacking=0`) are intended strictly for debugging purposes and should not be used in production environments. Therefore, until a proper fix is released, the safest option for production is to downgrade to version 26.05.

Sorry for any confusion my earlier comment may have caused. We will update you here as soon as the fix is ready.

**Comment by encodatamHirmer** [comment]
I can reproduce what looks like the same issue on a system with 2× Intel Arc Pro B60, Ubuntu 26.04, compute-runtime 26.27.39122.12, PyTorch 2.13.0+xpu and vLLM 0.27.0.

With kernel 7.0.0-29-generic, vLLM tensor parallelism (TP=2) reproducibly ends in host memory exhaustion. The kernel log shows:

xe_ttm_tt_populate -> xe_gem_prime_export -> drm_gem_prime_handle_to_dmabuf

followed by global OOM killer activity. vLLM itself reports XPU OOM even though several GiB of VRAM are still free. For example: Tried to allocate 962 MiB ... 9.06 GiB free

I then performed an A/B test while keeping the entire userspace unchanged:

Ubuntu 26.04
compute-runtime 26.27.39122.12
PyTorch 2.13.0+xpu
vLLM 0.27.0
same model and same vLLM command line

The only change was the kernel:

7.0.0-29-generic: fails reproducibly with host OOM
6.17.0-41-generic: works

Small XCCL collectives also succeed on the 7.0 kernel, so the failure only becomes visible with the larger real-world allocations performed by vLLM. This seems to indicate that the issue depends on the interaction between compute-runtime and the newer xe kernel driver / DMA-BUF peer-memory path, rather than compute-runtime alone.

I am currently using kernel 6.17 as a workaround.

Please let me know if additional traces or a minimal reproducer would be useful.

This may also be related to #952, since that issue also involves dual Arc Pro B60 systems where small synthetic tests work while larger real workloads fail.

AI-assisted wording; all technical observations and test results are my own.

**Comment by jwarchul** [comment]
Hi @blanelegant, @encodatamHirmer,

This issue should now be resolved in the latest release: [26.35.39758.10](https://github.com/intel/compute-runtime/releases/tag/26.35.39758.10).
Could you please confirm if it works on your end?

Thank you for your patience and cooperation.

**Comment by atorax** [comment]
Independent confirmation on 2× Arc Pro B70 (Battlemage G31), and some additional datapoints — this regression is still present in **26.31.39395.13**, and it does not require DDP or any multi-device workload to trigger: single-process, single-device allocations are enough as long as a second GPU is *visible*.

**Environment:** Debian 13, kernel 7.1.8+deb13-amd64 (xe), `intel-opencl-icd`/`libze-intel-gpu1` 26.31.39395.13, `libze1` 1.33.1, torch 2.14.0+xpu, host 64GB RAM.

**Minimal repro** (30 lines, bare `torch.empty()` on `xpu:0`, no frameworks): full script and results in intel/torch-xpu-ops#5428. Summary of `/proc/meminfo` deltas:

| condition | 16GiB allocated on xpu:0 | host effect |
|---|---|---|
| both GPUs visible | MemFree −5.1GiB per step, swap absorbing the rest; Committed_AS +16.3GiB | 1:1 host pinning |
| `ZE_AFFINITY_MASK=0` (one GPU visible) | **MemFree −0.1GiB**; Committed_AS still +16.3GiB | pinning gone |

The mask on/off toggle on the same binary/box points squarely at the peer-registration pinning you describe: the commit accounting exists either way, but physical host pages are only pinned when a peer device is visible.

Additional datapoints:
- Also reproduces identically on torch **2.11.0+xpu** (IPEX lineage) — allocator-agnostic within the torch stack.
- llama.cpp (SYCL and Vulkan) on the same host fills both cards ~55GB with **no** host pinning — consistent with it never establishing cross-device peer access.
- At production scale (45GB of model weights split across both cards via PyTorch), the pinned mirror consumes essentially all host RAM on a 64GB machine; since pinned pages are unswappable, the kernel evicts everything else and then OOM-kills the workload. Reproduced 6/6.

Happy to test candidate fixes or debug keys on this hardware.

**Comment by jwarchul** [comment]
Hi @atorax,

The [26.31.39395.13](https://github.com/intel/compute-runtime/releases/tag/26.31.39395.13) release does not include this fix yet. Could you please test with the [26.35.39758.10](https://github.com/intel/compute-runtime/releases/tag/26.35.39758.10) release instead?

Thank you.

**Comment by atorax** [comment]
Tested 26.35.39758.10 (with matched IGC 2.41.5, gmmlib 22.10.0) on the same 2× Arc Pro B70 / Debian 13 / torch 2.14.0+xpu setup — **confirmed fixed**.

- Bare-tensor repro (both GPUs visible, 8GiB on xpu:0): MemAvailable moves by noise only (~0.1GiB). Notably, `Committed_AS` no longer tracks device allocations 1:1 at all — on 26.05 the commit accounting was still present (just unpinned), so 26.35 appears to remove the host mirror entirely rather than merely not pinning it.
- Production scale: 45GB of model weights split across both cards via PyTorch (FLUX.2-dev), full render completes with host RAM fully recovered while weights stay device-resident; wall time at parity with 26.31 (190s vs 197s). Model residency across renders — impossible under the regression on a 64GB host — works again.

Thanks for the fast turnaround.

**Comment by encodatamHirmer** [comment]
Hi @jwarchul 
I tested again with the complete compute-runtime **26.35.39758.10** stack and now have a clean kernel A/B comparison.

Installed userspace stack in both tests:

```text
intel-opencl-icd     26.35.39758.10-0
libze-intel-gpu1     26.35.39758.10-0
intel-ocloc          26.35.39758.10-0
intel-igc-core-2     2.41.5
intel-igc-opencl-2   2.41.5
libigdgmm12          22.10.0
```

Hardware:

```text
2x Intel Arc Pro B60 24 GB
32 GB system RAM
vLLM using tensor parallelism across both GPUs
```

With Linux **7.0.0-31-generic**, the problem is still reproducible.

vLLM successfully loads the model and finishes profiling. It then reports approximately 9 GiB of available KV-cache memory per GPU:

```text
Available KV cache memory: 9.0 GiB
GPU KV cache size: 516,840 tokens
```

During KV-cache initialization, host memory is rapidly exhausted and a TP worker is killed.

The kernel reports a global OOM:

```text
oom-kill: ... global_oom
Out of memory: Killed process ... (VLLM::Worker_TP)
```

The failed allocation is in the xe PRIME/dma-buf path:

```text
VLLM::Worker_TP: page allocation failure: order:0
xe_ttm_tt_populate
xe_gem_prime_export
drm_gem_prime_handle_to_dmabuf
drm_prime_handle_to_fd_ioctl
xe_drm_ioctl
```

I monitored host memory every 200 ms. Immediately before the crash, `MemAvailable` dropped to approximately **200 MB** on the 32 GB host.

When the TP workers were killed, large amounts of host memory were released immediately and `MemAvailable` returned to approximately **28 GB**.

I then rebooted the same machine into:

```text
6.17.0-41-generic
```

and repeated the test with **exactly the same compute-runtime 26.35 stack, container and vLLM configuration**.

With kernel 6.17, the workload starts successfully.

During the same KV-cache initialization phase, host memory behaves completely differently. Memory usage rises only by a few GB and `MemAvailable` remains around **15–19 GB**. vLLM completes initialization and reaches the ready state without an OOM.

So the result is:

```text
compute-runtime 26.35.39758.10 + Linux 7.0.0-31-generic -> global host OOM
compute-runtime 26.35.39758.10 + Linux 6.17.0-41-generic -> works
```

This strongly suggests that the remaining/reproduced behavior depends on the kernel/xe driver path introduced or changed between these kernel versions.

It may therefore not be exactly the original userspace regression anymore, but the observable behavior is very similar: large multi-GPU allocations result in massive host-memory consumption, and the failing kernel stack goes through `xe_ttm_tt_populate()` while exporting the GEM object through PRIME/dma-buf.

This comment was drafted with the assistance of AI, but I reviewed and verified the technical content before posting it.

**Comment by jwarchul** [comment]
@atorax, thanks for testing and verifying the fix.

@encodatamHirmer, this seems to be a separate issue from what was originally reported in this ticket. Could you please open a new GitHub issue for this? Please provide a minimal reproducer and test if running with `NEOReadDebugKeys=1 EnableDeferBacking=0` helps on your end. We would really appreciate it.
